crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4301,6 +4402,12 @@ impl<'db> TypeInferenceBuilder<'db> {
         let mut call_arguments = Self::parse_arguments(arguments);
         let callable_type = self.infer_expression(func);
 
+        if let Type::FunctionLiteral(function) = callable_type {
+            if function.definition(self.db()).scope(self.db()) == self.scope() {
```

Comment:
This line is the source of the test failure (and probably most of the regression, too.) If you look at the body of `FunctionType::definition`, you'll see that it accesses the semantic index, which means it has a direct dependency on the AST of the file the function is defined in. We are calling this unconditionally on every `FunctionLiteral` called, including those imported from other modules, so this will create direct dependencies to the ASTs of those modules.

One way we can prevent this is to first check that `function.body_scope(self.db()).file(self.db()) == self.file()` and only if that is true, go ahead and get the definition and do the full scope-match check.

Reply:
Thank you, this was the case. I've added a warning note as well.